session.go

```diff
@@ -157,6 +157,16 @@ func (wac *Conn) Login(qrChan chan<- string, ctx context.Context) (Session, JID,
 
 	wac.loginSessionLock.Lock()
 	defer wac.loginSessionLock.Unlock()
+
+	defer func() {
```

Comment:
Dumb `go` question? Why does this sit inside a defer rather than outside? Does the defer somehow protect us from nasties?

Reply:
`defer` is a "run this when this function finishes, for whatever reason". So we'll get this if it finishes normally (that'd make `wac.loggedIn` true and so we'd skip this), prematurely due to timeouts and such (which will `Disconnect()`) or crashes entirely (which will still `Disconnect` though it won't matter at that point).

It's effectively a `finally` for the surrounding function.

Reply:
That's pretty cool, and I definitely didn't expect that behaviour!

Reply:
I'm generally unsure if code doing unexpected things is cool :grin: 

/me should also learn more Go.